File: clients/android/core/src/main/kotlin/com/migo/core/domain/Keys.kt
package com.migo.core.domain

import com.migo.core.crypto.IdentityPublic
import com.migo.core.crypto.IdentitySecret
import com.migo.core.crypto.KeyPair
import com.migo.core.crypto.OneTimePrekey
import com.migo.core.crypto.PrekeyBundle
import com.migo.core.crypto.SignedPrekey
import com.migo.core.protocol.KeyBundle
import com.migo.core.protocol.KeyBundleRequest
import com.migo.core.protocol.KeyBundleResponse
import com.migo.core.protocol.KeyPublish
import com.migo.core.protocol.KeyPublishResult
import com.migo.core.protocol.Op
import com.migo.core.protocol.PrekeyEntry
import com.migo.core.session.IdentityProvider
import com.migo.core.session.LocalKeyStore
import com.migo.core.session.PeerBundleSource
import com.migo.core.store.DeviceKeys
import com.migo.core.store.SavedSession
import com.migo.core.wire.Id
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

/**
 * Key material and the key-directory domain.
 *
 * Two things live here, mirroring `packages/sdk/src/domains/keys.ts`. [KeyStore] is this device's
 * private key material -- the identity, the current signed prekey, and the pool of one-time prekeys.
 * [KeysDomain] is the wire side: it publishes the *public* halves via `KEY_PUBLISH` and fetches peers'
 * bundles via `KEY_BUNDLE_FETCH`.
 *
 * The store satisfies both crypto-layer seams: [LocalKeyStore], which is what the 1:1 layer needs to
 * answer a first message (resolve a signed or one-time prekey by id), and [IdentityProvider], which is
 * what the group layer needs in order to sign a sender key. One object backs both, so a client wires a
 * single [KeyStore] into [com.migo.core.session.SessionCrypto] and
 * [com.migo.core.session.GroupCrypto] alike, and the two layers cannot end up signing with different
 * identities.
 *
 * # What crosses the wire
 *
 * Only public keys, key ids, and a signature. [KeyStore.publish] builds a [KeyPublish] out of public
 * halves; the seeds behind them stay inside the crypto objects, which hand out copies and never the
 * live buffers. That is the invariant making the server untrusted for confidentiality: it can choose
 * *which* bundle to serve, but every bundle it serves is verified against the claimed identity's
 * signature on the fetching device before any Diffie-Hellman happens -- [SignedPrekey.verify], called
 * by [com.migo.core.crypto.X3dh.initiate].
 *
 * # Why this file locks and the reference does not
 *
 * The web SDK runs on one event loop, so its store needs no synchronisation. Here the same store is
 * reached from two crypto objects that hold *different* locks: `SessionCrypto` guards its sessions
 * with a coroutine `Mutex`, `GroupCrypto` with a `ReentrantLock`, and neither knows about the other.
 * Two coroutines on different dispatchers can therefore be inside [consumeOneTimePrekey] and
 * [replenishOneTimePrekeys] at the same instant, which on a plain `HashMap` is not a lost entry but a
 * corrupted table. The lock here is what makes the store safe to share, and it is a plain
 * [ReentrantLock] rather than a `Mutex` because [LocalKeyStore] is a non-suspending interface -- it is
 * called from inside `SessionCrypto`'s own critical section, where suspending is not available.
 *
 * # No separate snapshot type
 *
 * The reference exports a `KeyStoreSnapshot` because a web caller persists the seeds itself. This
 * client already has exactly one place private key material is written -- [DeviceKeys], sealed by
 * [com.migo.core.store.Vault] under an Android Keystore key -- so [DeviceKeys] *is* the snapshot, and
 * [restore] and [export] are the two directions. A second representation of the same secret state
 * would be a second thing to keep in step, and the failure when it drifted would be a device that
 * silently loses its identity on restart.
 */

/** How many one-time prekeys a fresh store mints, matching the server's expected batch size. */
const val DEFAULT_ONE_TIME_PREKEYS = 64

/**
 * The most one-time prekeys the pool may hold at once.
 *
 * A hundred, because that is `migo_keys::model::MAX_ONE_TIME_PREKEYS` and the server refuses a
 * publication carrying more with `VALIDATION_FAILED`. [KeyStore.publish] publishes the *whole* pool
 * and publication replaces rather than merges, so a pool over the cap is not a partial publish: it is
 * a device that cannot publish at all until it drops keys. [KeyStore.replenishOneTimePrekeys]
 * enforces the bound so a caller cannot construct that state by asking twice.
 */
const val MAX_ONE_TIME_PREKEYS = 100

/**
 * The largest key id this client will mint.
 *
 * The server narrows every key id from `u32` to `i32` (`migo-keys/src/service.rs`, `fn key_id`) and
 * refuses anything that does not fit, so the wire's `u32` is not the real bound. Reaching this would
 * take four billion rotations on one device; the check exists so that if it ever did happen the
 * failure would be this local error rather than a publication the server rejects for reasons the
 * client cannot see.
 */
private const val MAX_KEY_ID = 2_147_483_647L

/** One published signed prekey: the id, the pair behind it, and the identity's signature over it. */
private class SignedPrekeyEntry(
    val keyId: Long,
    val pair: KeyPair,
    val signed: SignedPrekey,
)

/**
 * This device's private key material.
 *
 * Construct one with [KeyStore.create] for a new device, or [KeyStore.restore] to rebuild from the
 * [DeviceKeys] the vault held. Everything the two crypto layers ask of local material is answered
 * here; nothing here leaves the device except through [publish], which emits public data only.
 */
class KeyStore private constructor(
    private val identityKey: IdentitySecret,
    private var signedPrekey: SignedPrekeyEntry,
    private var nextSignedPrekeyId: Long,
    private var nextOneTimePrekeyId: Long,
) : LocalKeyStore, IdentityProvider {

    private val lock = ReentrantLock()
    private val oneTimePrekeys = HashMap<Long, KeyPair>()

    /**
     * This device's long-term identity secret. Backs both crypto layers.
     *
     * Returned rather than copied, because [IdentitySecret] is the holder of the seeds and there is
     * nothing to copy that would not be a second holder of the same secret. It is immutable, so
     * sharing it is safe; no lock is taken because the field never changes after construction.
     */
    override fun identity(): IdentitySecret = identityKey

    /** The pair for a published signed prekey id, or null once that id has been rotated away. */
    override fun signedPrekeyPair(signedPrekeyId: Long): KeyPair? = lock.withLock {
        if (signedPrekey.keyId == signedPrekeyId) signedPrekey.pair else null
    }

    /**
     * The pair for a published one-time prekey id *without* consuming it, or null if we do not hold
     * it.
     *
     * Peeking rather than consuming is what lets the 1:1 layer attempt a responder handshake for a
     * first message that was broadcast to every device and only spend the prekey once the decrypt
     * proves the message was for this one. Consuming here would burn a prekey on every sibling
     * device's copy of somebody else's message.
     */
    override fun oneTimePrekeyPair(keyId: Long): KeyPair? = lock.withLock {
        oneTimePrekeys[keyId]
    }

    /**
     * Permanently removes a one-time prekey from the pool, after a first message using it has opened.
     *
     * Idempotent: consuming an id that is already gone is a no-op, so a lost race cannot throw. A
     * replayed first message finds the prekey gone and can never derive a second session from it,
     * which is the property the one-time prekey exists for.
     */
    override fun consumeOneTimePrekey(keyId: Long) {
        lock.withLock { oneTimePrekeys.remove(keyId) }
    }

    /** How many unused one-time prekeys remain, so the client knows when to replenish. */
    fun oneTimePrekeyCount(): Int = lock.withLock { oneTimePrekeys.size }

    /**
     * Adds up to [count] fresh one-time prekeys and returns the public entries that were minted.
     *
     * Clamped to the headroom under [MAX_ONE_TIME_PREKEYS], and it returns what it actually minted
     * rather than what was asked for. Clamping instead of throwing is the right failure here: a
     * caller that asked for more than fits still ends up with a full pool and a publication the
     * server accepts, where a throw would leave a client unable to top up at all once its policy and
     * the server's cap disagreed.
     */
    fun replenishOneTimePrekeys(count: Int): List<PrekeyEntry> = lock.withLock {
        val room = MAX_ONE_TIME_PREKEYS - oneTimePrekeys.size
        val minting = minOf(count, room).coerceAtLeast(0)
        val added = ArrayList<PrekeyEntry>(minting)
        for (i in 0 until minting) {
            val keyId = takeOneTimePrekeyIdLocked()
            val pair = KeyPair.generate()
            oneTimePrekeys[keyId] = pair
            added.add(PrekeyEntry(keyId, pair.public()))
        }
        added
    }

    /**
     * Rotates in a new signed prekey, retiring the old one. The client republishes afterward.
     *
     * The retired pair is dropped, which is deliberate and is what forward secrecy for the signed
     * prekey means: a device seized tomorrow cannot answer a first message composed against the
     * prekey it published last month. The cost is that such a message no longer opens, which is why
     * section 163 gives the prekey a thirty-day lifetime rather than rotating it per message.
     */
    fun rotateSignedPrekey() {
        lock.withLock {
            val keyId = nextSignedPrekeyId
            if (keyId > MAX_KEY_ID) throw SdkError("keys: signed prekey ids exhausted")
            nextSignedPrekeyId = keyId + 1L
            signedPrekey = buildSignedPrekey(identityKey, keyId)
        }
    }

    /**
     * The public key material to publish to the server.
     *
     * The identity goes as its 64-byte wire form; the signed prekey carries the signature binding it
     * to that identity; every one-time prekey currently held is offered, because publication replaces
     * what the server has and a partial list would retire prekeys this device can still answer for.
     */
    fun publish(): KeyPublish = lock.withLock {
        val entries = ArrayList<PrekeyEntry>(oneTimePrekeys.size)
        for ((keyId, pair) in oneTimePrekeys) {
            entries.add(PrekeyEntry(keyId, pair.public()))
        }
        KeyPublish(
            identityKey = identityKey.public().toBytes(),
            signedPrekeyId = signedPrekey.keyId,
            signedPrekey = signedPrekey.signed.publicKey,
            signedPrekeySignature = signedPrekey.signed.signature,
            oneTimePrekeys = entries,
        )
    }

    /** The public identity, for fingerprint display and contact-change detection. */
    fun publicIdentity(): IdentityPublic = identityKey.public()

    /**
     * The full private state, for [com.migo.core.store.Vault] to seal.
     *
     * [session] is threaded through rather than held here because a sign-in is not key material: the
     * vault stores both in one blob, but the store has no business knowing a refresh token. The
     * returned map is a fresh copy, so a save that happens while a prekey is being consumed writes a
     * coherent pool rather than one being mutated underneath it.
     */
    fun export(session: SavedSession?): DeviceKeys = lock.withLock {
        DeviceKeys(
            identity = identityKey,
            signedPrekeyId = signedPrekey.keyId,
            signedPrekey = signedPrekey.pair,
            oneTime = HashMap(oneTimePrekeys),
            nextSignedPrekeyId = nextSignedPrekeyId,
            nextOneTimePrekeyId = nextOneTimePrekeyId,
            session = session,
        )
    }

    /** Public shape only. Never a seed, and never a count that is not already public. */
    override fun toString(): String =
        "KeyStore(identity: ${identityKey.public()}, signed_prekey_id: ${signedPrekey.keyId}, " +
            "one_time_count: ${oneTimePrekeyCount()})"

    /** Reserves the next one-time prekey id. Caller holds [lock]. */
    private fun takeOneTimePrekeyIdLocked(): Long {
        val keyId = nextOneTimePrekeyId
        if (keyId > MAX_KEY_ID) throw SdkError("keys: one-time prekey ids exhausted")
        nextOneTimePrekeyId = keyId + 1L
        return keyId
    }

    companion object {
        /**
         * Mints a new device's key material: a fresh identity, one signed prekey, and a batch of
         * one-time prekeys.
         *
         * The caller persists the result ([export], through the vault) and publishes the public
         * halves ([KeysDomain.publish]). Both matter: an identity that was generated but never saved
         * is a device that changes its safety number on every launch, and one that was saved but
         * never published is a device nobody can start a conversation with.
         */
        fun create(oneTimePrekeyCount: Int = DEFAULT_ONE_TIME_PREKEYS): KeyStore {
            val identity = IdentitySecret.generate()
            val store = KeyStore(identity, buildSignedPrekey(identity, 1L), 2L, 1L)
            store.replenishOneTimePrekeys(oneTimePrekeyCount)
            return store
        }

        /**
         * Rebuilds a store from the [DeviceKeys] the vault held.
         *
         * The signed prekey's signature is recomputed over the restored pair rather than stored,
         * because a signature is derivable from material already in the vault and storing it would
         * be one more thing that could be inconsistent with the key it signs. A restore therefore
         * round-trips to a byte-identical publication.
         */
        fun restore(keys: DeviceKeys): KeyStore {
            val entry = SignedPrekeyEntry(
                keys.signedPrekeyId,
                keys.signedPrekey,
                SignedPrekey.create(keys.identity, keys.signedPrekeyId, keys.signedPrekey),
            )
            val store = KeyStore(
                keys.identity,
                entry,
                keys.nextSignedPrekeyId,
                keys.nextOneTimePrekeyId,
            )
            store.oneTimePrekeys.putAll(keys.oneTime)
            return store
        }
    }
}

/** Rebuilds a signed prekey entry for [keyId] from a fresh pair signed by [identity]. */
private fun buildSignedPrekey(identity: IdentitySecret, keyId: Long): SignedPrekeyEntry {
    val pair = KeyPair.generate()
    return SignedPrekeyEntry(keyId, pair, SignedPrekey.create(identity, keyId, pair))
}

/** A peer device's id paired with its bundle, for enumerating a user's devices. */
class DeviceBundle(
    /** The device the bundle belongs to. */
    val deviceId: Id,
    /** Its published key material, rebuilt into the crypto layer's type. */
    val bundle: PrekeyBundle,
) {
    /** Public key material only; safe to log. */
    override fun toString(): String = "DeviceBundle(device_id: $deviceId)"
}

/**
 * The key-directory domain: publish our public keys, fetch peers'.
 *
 * Implements [PeerBundleSource] so the 1:1 layer can fetch one device's bundle on demand the first
 * time it needs to become an initiator. Also exposes [fetchDeviceBundles], which the messaging layer
 * needs because a sender key has to be distributed to every one of a user's devices, not to a user.
 */
class KeysDomain(
    private val rpc: Rpc,
    private val store: KeyStore,
) : PeerBundleSource {

    /**
     * Publishes this device's current public key material.
     *
     * Call it after [KeyStore.create], after [KeyStore.rotateSignedPrekey], and after
     * [KeyStore.replenishOneTimePrekeys]. A replenish that is not followed by a publish leaves the
     * server handing out the old batch, and every session formed from one of those is undecryptable
     * by this device.
     */
    suspend fun publish(): KeyPublishResult {
        val request = store.publish()
        return rpc.call(Op.KEY_PUBLISH, { w -> request.encode(w) }, { r -> KeyPublishResult.decode(r) })
    }

    /**
     * Fetches the bundle for one device of one user.
     *
     * Fails with [SdkError] when the server returns no bundle for the device that was asked for,
     * which covers both an empty response and a response about somebody else's devices. It is a
     * domain failure rather than a transport one: the connection is fine and exactly this operation
     * cannot proceed.
     *
     * The bundle is deliberately *not* verified here. [com.migo.core.crypto.X3dh.initiate] verifies
     * it immediately before the key agreement, and keeping verification in that one place is what
     * makes it impossible to reach a Diffie-Hellman through a path that forgot to check. Verifying
     * twice would read as a belt-and-braces improvement and would in fact make the single mandatory
     * check look optional.
     */
    override suspend fun fetchBundle(userId: Id, deviceId: Id): PrekeyBundle {
        val request = KeyBundleRequest(userId, deviceId)
        val response = rpc.call(
            Op.KEY_BUNDLE_FETCH,
            { w -> request.encode(w) },
            { r -> KeyBundleResponse.decode(r) },
        )
        val wire = response.bundles.firstOrNull { it.deviceId == deviceId }
            ?: throw SdkError("keys: server returned no bundle for device $deviceId")
        return toPrekeyBundle(wire)
    }

    /**
     * Fetches every device bundle a user currently publishes.
     *
     * The messaging layer uses this to learn which devices to seal a sender-key distribution for.
     * Each bundle returned consumes a one-time prekey on the server, so callers fetch once per
     * distribution round and never per message; the server caps one fetch at twenty devices
     * (`MAX_BUNDLES_PER_FETCH`) for the same reason.
     */
    suspend fun fetchDeviceBundles(userId: Id): List<DeviceBundle> {
        val request = KeyBundleRequest(userId)
        val response = rpc.call(
            Op.KEY_BUNDLE_FETCH,
            { w -> request.encode(w) },
            { r -> KeyBundleResponse.decode(r) },
        )
        return response.bundles.map { DeviceBundle(it.deviceId, toPrekeyBundle(it)) }
    }
}

/**
 * Rebuilds a crypto-layer [PrekeyBundle] from the wire [KeyBundle].
 *
 * A malformed identity key or a wrong-length prekey throws out of the constructors rather than being
 * repaired, and the throw is a [com.migo.core.crypto.CryptoError] that says which length was wrong
 * without carrying the bytes. A server that serves structurally broken material is a server whose
 * bundles this device must not use, and there is no partial bundle worth keeping.
 */
private fun toPrekeyBundle(wire: KeyBundle): PrekeyBundle {
    val identity = IdentityPublic.parse(wire.identityKey)
    val signedPrekey = SignedPrekey(
        wire.signedPrekeyId,
        wire.signedPrekey,
        wire.signedPrekeySignature,
    )
    // Both halves or neither. A key id with no key, or a key with no id, is a bundle the server
    // assembled wrongly, and treating it as "no one-time prekey" would silently drop this session
    // to the signed prekey alone -- a weaker guarantee, chosen by the untrusted party, with nothing
    // anywhere saying it happened.
    val keyId = wire.oneTimePrekeyId
    val publicKey = wire.oneTimePrekey
    val oneTimePrekey = when {
        keyId != null && publicKey != null -> OneTimePrekey(keyId, publicKey)
        keyId == null && publicKey == null -> null
        else -> throw SdkError("keys: bundle carries half a one-time prekey")
    }
    return PrekeyBundle(identity, signedPrekey, oneTimePrekey)
}

File: packages/sdk/src/domains/keys.ts
/**
 * Key material and the key-directory domain.
 *
 * Two things live here. {@link KeyStore} is this device's private key material — the identity, the
 * current signed prekey, and the pool of one-time prekeys — held in memory and never serialised to
 * anything but the caller's own secure storage. {@link KeysDomain} is the wire side: it publishes the
 * *public* halves via KEY_PUBLISH and fetches peers' bundles via KEY_BUNDLE_FETCH.
 *
 * The store deliberately satisfies both crypto-layer seams: {@link LocalKeyStore} (what the 1:1 layer
 * needs to answer a first message — resolve a signed or one-time prekey by id) and
 * {@link IdentityProvider} (what the group layer needs to sign). One object backs both, so the client
 * wires a single {@link KeyStore} into {@link SessionCrypto} and {@link GroupCrypto} alike.
 *
 * # What crosses the wire
 *
 * Only public keys, key ids, and a signature. {@link KeyStore.publish} builds a {@link KeyPublish}
 * from public halves; the seeds behind them stay in the private fields of the crypto objects. This is
 * the invariant that makes the server untrusted for confidentiality: it can pick which bundle to
 * serve, but every bundle it serves is verified against the claimed identity's signature on the
 * fetching device before any Diffie-Hellman ({@link SignedPrekey.verify}, called by {@link initiate}).
 */

import type { Id } from '@migo/wire';
import {
  account,
  IdentityPublic,
  IdentitySecret,
  KeyPair,
  PrekeyBundle,
  SignedPrekey,
} from '@migo/crypto';
import {
  OP,
  encodeKeyBundleRequest,
  encodeKeyPublish,
  decodeKeyBundleResponse,
  decodeKeyPublishResult,
} from '@migo/protocol';
import type { KeyBundle, KeyPublish, KeyPublishResult, PrekeyEntry } from '@migo/protocol';

import { SdkError } from '../errors.js';
import type { IdentityProvider } from '../group-crypto.js';
import type { LocalKeyStore, PeerBundleSource } from '../session-crypto.js';
import type { Rpc } from './rpc.js';

/** How many one-time prekeys a fresh store mints, matching the server's expected batch size. */
const DEFAULT_ONE_TIME_PREKEYS = 64;

/** One published signed prekey: the id, the pair behind it, and the identity's signature over it. */
interface SignedPrekeyEntry {
  keyId: number;
  pair: KeyPair;
  signed: SignedPrekey;
}

/**
 * A serialisable snapshot of a {@link KeyStore}: every private seed plus the id counters.
 *
 * This is raw secret material. It exists so a device can persist its identity to the caller's own
 * secure local storage and rebuild the store after a restart without re-registering — it is never
 * transmitted anywhere. Seeds are 32 bytes each ({@link SEED_LEN}); the store reconstructs the full
 * key pairs from them, so signatures and public keys never need to be stored.
 */
export interface KeyStoreSnapshot {
  identitySigningSeed: Uint8Array;
  identityExchangeSeed: Uint8Array;
  signedPrekeyId: number;
  signedPrekeySeed: Uint8Array;
  oneTimePrekeys: Array<{ keyId: number; seed: Uint8Array }>;
  nextSignedPrekeyId: number;
  nextOneTimePrekeyId: number;
  /**
   * The unified account root's 32 bytes, on the device that founded the account. Absent on every
   * additional device: they generate a fresh identity and never inherit the root, which is what
   * keeps the root the one secret a `.migo` container has to carry.
   */
  root?: Uint8Array;
  /** The tracked AVAX transactions, newest first. Absent rather than empty when there are none. */
  trackedTxs?: TrackedTx[];
}

/**
 * One tracked AVAX transaction: what was sent, and how the tracker ended (§184, spec #59).
 *
 * The chain has no "list transactions by sender" without an indexer, so the Activity list is a
 * client-side record. It rides the key-store snapshot because that is already the device's sealed
 * local state — nothing here is transmitted anywhere. Wei magnitudes are `bigint` by construction;
 * a `number` at a call site is a silent precision bug the type prevents.
 */
export interface TrackedTx {
  /** The transaction hash, the handle the chain knows it by. */
  txHash: Uint8Array;
  /** The chain the transaction was signed for — EIP-155's replay protection, restated. */
  chainId: number;
  /** The recipient. */
  to: Uint8Array;
  /** The amount, wei. AVAX has 18 decimals. */
  valueWei: bigint;
  /** The fee ceiling the user confirmed: `maxFeePerGas * gasLimit`, wei. */
  feeWei: bigint;
  /** The gas limit that was signed. */
  gasLimit: number;
  /** When the transaction was broadcast, unix seconds. */
  atUnix: number;
  /** Spec #41's own word for where the transaction stands: `PENDING` at broadcast, one of the
   *  tracker's endings once it settles. */
  outcome: string;
  /** The block that included the transaction, once one did. */
  block?: number;
  /** The gas the block actually spent on it, from the receipt — the ceiling's honest companion. */
  gasUsed?: bigint;
}

/**
 * This device's private key material.
 *
 * Construct one with {@link KeyStore.create} for a new device, or {@link KeyStore.restore} to rebuild
 * from a {@link KeyStoreSnapshot} the caller persisted. Everything the two crypto layers ask of local
 * material is answered here; nothing here is sent anywhere except through {@link KeyStore.publish},
 * which emits public data.
 */
export class KeyStore implements LocalKeyStore, IdentityProvider {
  readonly #identity: IdentitySecret;
  #signedPrekey: SignedPrekeyEntry;
  readonly #oneTimePrekeys = new Map<number, KeyPair>();
  #nextSignedPrekeyId: number;
  #nextOneTimePrekeyId: number;
  readonly #root: account.MigoRoot | null;
  readonly #trackedTxs: TrackedTx[];

  private constructor(
    identity: IdentitySecret,
    signedPrekey: SignedPrekeyEntry,
    nextSignedPrekeyId: number,
    nextOneTimePrekeyId: number,
    root: account.MigoRoot | null,
    trackedTxs: TrackedTx[],
  ) {
    this.#identity = identity;
    this.#signedPrekey = signedPrekey;
    this.#nextSignedPrekeyId = nextSignedPrekeyId;
    this.#nextOneTimePrekeyId = nextOneTimePrekeyId;
    this.#root = root;
    this.#trackedTxs = trackedTxs;
  }

  /**
   * Mints a new device's key material: a fresh identity, one signed prekey, and a batch of one-time
   * prekeys. The caller persists the seeds ({@link KeyStore.snapshot}) and publishes the public
   * halves ({@link KeysDomain.publish}).
   */
  static create(oneTimePrekeyCount: number = DEFAULT_ONE_TIME_PREKEYS): KeyStore {
    const identity = IdentitySecret.generate();
    const signedPrekey = buildSignedPrekey(identity, 1);
    const store = new KeyStore(identity, signedPrekey, 2, 1, null, []);
    store.replenishOneTimePrekeys(oneTimePrekeyCount);
    return store;
  }

  /**
   * Mints the founding device of a brand-new account (§182): the E2EE identity is derived from the
   * root's E2EE domain rather than generated, which is what makes the account's E2EE history
   * recoverable from a `.migo` container. Additional devices never take this path — they call
   * {@link KeyStore.create} and hold no root.
   */
  static founding(
    root: account.MigoRoot,
    oneTimePrekeyCount: number = DEFAULT_ONE_TIME_PREKEYS,
  ): KeyStore {
    const seeds = account.foundingDeviceE2eeSeeds(root);
    const identity = IdentitySecret.fromSeeds(seeds.signing, seeds.exchange);
    const signedPrekey = buildSignedPrekey(identity, 1);
    const store = new KeyStore(identity, signedPrekey, 2, 1, root, []);
    store.replenishOneTimePrekeys(oneTimePrekeyCount);
    return store;
  }

  /**
   * Rebuilds a store from a {@link KeyStoreSnapshot} produced by {@link KeyStore.snapshot}.
   *
   * Key pairs are reconstructed from their seeds, and the signed prekey's signature is recomputed
   * over the restored pair, so a snapshot round-trips to a byte-identical published bundle. The
   * root and the tracked transactions ride along when the snapshot carries them, which is only on
   * the device that founded the account.
   */
  static restore(snapshot: KeyStoreSnapshot): KeyStore {
    const identity = IdentitySecret.fromSeeds(
      snapshot.identitySigningSeed,
      snapshot.identityExchangeSeed,
    );
    const signedPrekeyPair = KeyPair.fromSeed(snapshot.signedPrekeySeed);
    const signedPrekey: SignedPrekeyEntry = {
      keyId: snapshot.signedPrekeyId,
      pair: signedPrekeyPair,
      signed: SignedPrekey.create(identity, snapshot.signedPrekeyId, signedPrekeyPair),
    };
    const root = snapshot.root !== undefined ? account.MigoRoot.fromBytes(snapshot.root) : null;
    const store = new KeyStore(
      identity,
      signedPrekey,
      snapshot.nextSignedPrekeyId,
      snapshot.nextOneTimePrekeyId,
      root,
      snapshot.trackedTxs !== undefined ? [...snapshot.trackedTxs] : [],
    );
    for (const entry of snapshot.oneTimePrekeys) {
      store.#oneTimePrekeys.set(entry.keyId, KeyPair.fromSeed(entry.seed));
    }
    return store;
  }

  /**
   * The unified account root, on the device that founded the account, or `null` on every additional
   * device. This is the one secret the wallet and the `.migo` container both derive from — hand it
   * to nothing but the `account` surface's own functions.
   */
  root(): account.MigoRoot | null {
    return this.#root;
  }

  /**
   * The tracked AVAX transactions, newest first — the wallet surface's live list.
   *
   * The array returned is the store's own: the wallet flow mutates it (a record inserted at
   * broadcast, its ending written at settle) and the next {@link snapshot} seals the result, which
   * is the same trade the one-time prekey pool makes.
   */
  trackedTxs(): TrackedTx[] {
    return this.#trackedTxs;
  }

  /** This device's long-term identity secret. Backs both crypto layers. */
  identity(): IdentitySecret {
    return this.#identity;
  }

  /** The pair for a published signed prekey id, or `null` once that id has been rotated away. */
  signedPrekeyPair(signedPrekeyId: number): KeyPair | null {
    return this.#signedPrekey.keyId === signedPrekeyId ? this.#signedPrekey.pair : null;
  }

  /**
   * The pair for a published one-time prekey id *without* consuming it, or `null` if we do not hold
   * it.
   *
   * Peeking rather than consuming is what lets the 1:1 layer attempt a responder handshake for a
   * broadcast first message and only spend the prekey once the decrypt proves the message was ours;
   * see {@link LocalKeyStore.oneTimePrekeyPair}.
   */
  oneTimePrekeyPair(keyId: number): KeyPair | null {
    return this.#oneTimePrekeys.get(keyId) ?? null;
  }

  /**
   * Permanently removes a one-time prekey from the pool, after a first message using it has opened.
   *
   * Idempotent: consuming an id that is already gone is a no-op, so a lost race cannot throw. A
   * replayed first message finds the prekey gone and can never derive a second session from it.
   */
  consumeOneTimePrekey(keyId: number): void {
    this.#oneTimePrekeys.delete(keyId);
  }

  /** How many unused one-time prekeys remain, so the client knows when to replenish. */
  oneTimePrekeyCount(): number {
    return this.#oneTimePrekeys.size;
  }

  /** Adds `count` fresh one-time prekeys to the pool and returns their public entries to publish. */
  replenishOneTimePrekeys(count: number): PrekeyEntry[] {
    const added: PrekeyEntry[] = [];
    for (let i = 0; i < count; i += 1) {
      const keyId = this.#nextOneTimePrekeyId;
      this.#nextOneTimePrekeyId += 1;
      const pair = KeyPair.generate();
      this.#oneTimePrekeys.set(keyId, pair);
      added.push({ keyId, publicKey: pair.public() });
    }
    return added;
  }

  /** Rotates in a new signed prekey, retiring the old one. The client republishes afterward. */
  rotateSignedPrekey(): void {
    const keyId = this.#nextSignedPrekeyId;
    this.#nextSignedPrekeyId += 1;
    this.#signedPrekey = buildSignedPrekey(this.#identity, keyId);
  }

  /**
   * The public key material to publish to the server.
   *
   * The identity is its 64-byte wire form; the signed prekey carries the signature that binds it to
   * the identity; every current one-time prekey is offered so peers can start forward-secret sessions.
   */
  publish(): KeyPublish {
    const oneTimePrekeys: PrekeyEntry[] = [];
    for (const [keyId, pair] of this.#oneTimePrekeys) {
      oneTimePrekeys.push({ keyId, publicKey: pair.public() });
    }
    return {
      identityKey: this.#identity.public().toBytes(),
      signedPrekeyId: this.#signedPrekey.keyId,
      signedPrekey: this.#signedPrekey.signed.publicKey,
      signedPrekeySignature: this.#signedPrekey.signed.signature,
      oneTimePrekeys,
    };
  }

  /** The public identity, for fingerprint display and contact-change detection. */
  publicIdentity(): IdentityPublic {
    return this.#identity.public();
  }

  /**
   * A snapshot of every private seed and id counter, for the caller to persist to secure local
   * storage. Never transmit this — it is the device's full secret state.
   */
  snapshot(): KeyStoreSnapshot {
    const oneTimePrekeys: Array<{ keyId: number; seed: Uint8Array }> = [];
    for (const [keyId, pair] of this.#oneTimePrekeys) {
      oneTimePrekeys.push({ keyId, seed: pair.exposeSeed() });
    }
    return {
      identitySigningSeed: this.#identity.exposeSigningSeed(),
      identityExchangeSeed: this.#identity.exposeExchangeSeed(),
      signedPrekeyId: this.#signedPrekey.keyId,
      signedPrekeySeed: this.#signedPrekey.pair.exposeSeed(),
      oneTimePrekeys,
      nextSignedPrekeyId: this.#nextSignedPrekeyId,
      nextOneTimePrekeyId: this.#nextOneTimePrekeyId,
      // The root is present only on a device that holds it, and the tx list only when it has
      // entries: a field that exists to say "nothing here" costs every reader a skip for no
      // information.
      ...(this.#root !== null ? { root: this.#root.asBytes() } : {}),
      ...(this.#trackedTxs.length > 0 ? { trackedTxs: [...this.#trackedTxs] } : {}),
    };
  }
}

/** Rebuilds a signed prekey entry for `keyId` from a fresh pair signed by `identity`. */
function buildSignedPrekey(identity: IdentitySecret, keyId: number): SignedPrekeyEntry {
  const pair = KeyPair.generate();
  return { keyId, pair, signed: SignedPrekey.create(identity, keyId, pair) };
}

/** A peer device's id paired with its verified bundle, for enumerating a user's devices. */
export interface DeviceBundle {
  deviceId: Id;
  bundle: PrekeyBundle;
}

/**
 * The key-directory domain: publish our public keys, fetch peers'.
 *
 * Implements {@link PeerBundleSource} so the 1:1 layer can fetch a single device's bundle on demand
 * when it first needs to become the initiator. Also exposes {@link fetchDeviceBundles} for the
 * messaging layer, which needs the *set* of a user's devices to distribute a sender key to each.
 */
export class KeysDomain implements PeerBundleSource {
  readonly #rpc: Rpc;
  readonly #store: KeyStore;

  constructor(rpc: Rpc, store: KeyStore) {
    this.#rpc = rpc;
    this.#store = store;
  }

  /** Publishes this device's current public key material. Call after create, rotate, or replenish. */
  async publish(): Promise<KeyPublishResult> {
    return this.#rpc.call(
      OP.KEY_PUBLISH,
      encodeKeyPublish,
      decodeKeyPublishResult,
      this.#store.publish(),
    );
  }

  /**
   * Fetches and verifies the bundle for one device of one user.
   *
   * Rejects if the server returns no bundle for the named device. The returned bundle is not yet
   * verified here — {@link initiate} verifies it before any key agreement, which is the single place
   * verification must happen so it cannot be forgotten.
   */
  async fetchBundle(userId: Id, deviceId: Id): Promise<PrekeyBundle> {
    const request = { userId, deviceId };
    const response = await this.#rpc.call(
      OP.KEY_BUNDLE_FETCH,
      encodeKeyBundleRequest,
      decodeKeyBundleResponse,
      request,
    );
    const wire = response.bundles.find((entry) => entry.deviceId === deviceId);
    if (wire === undefined) {
      throw new SdkError(`keys: server returned no bundle for device ${deviceId}`);
    }
    return toPrekeyBundle(wire);
  }

  /**
   * Fetches every device bundle a user currently publishes.
   *
   * The messaging layer uses this to learn which devices to seal a sender-key distribution for.
   * Fetching consumes one one-time prekey per device server-side, so callers fetch once per
   * distribution round, not per message.
   */
  async fetchDeviceBundles(userId: Id): Promise<DeviceBundle[]> {
    const request = { userId };
    const response = await this.#rpc.call(
      OP.KEY_BUNDLE_FETCH,
      encodeKeyBundleRequest,
      decodeKeyBundleResponse,
      request,
    );
    return response.bundles.map((wire) => ({
      deviceId: wire.deviceId,
      bundle: toPrekeyBundle(wire),
    }));
  }
}

/** Rebuilds a crypto-layer {@link PrekeyBundle} from the wire {@link KeyBundle}. */
function toPrekeyBundle(wire: KeyBundle): PrekeyBundle {
  const identity = IdentityPublic.parse(wire.identityKey);
  const signedPrekey = new SignedPrekey(
    wire.signedPrekeyId,
    wire.signedPrekey,
    wire.signedPrekeySignature,
  );
  const oneTimePrekey =
    wire.oneTimePrekeyId !== undefined && wire.oneTimePrekey !== undefined
      ? { keyId: wire.oneTimePrekeyId, publicKey: wire.oneTimePrekey }
      : null;
  return new PrekeyBundle(identity, signedPrekey, oneTimePrekey);
}
